Upload progress

jmcarp opened the pull request:
Add attachment upload progress, upload cancel / attachment delete, and draggable uploads. Still basically unstyled--if the behavior makes sense, I'll ask for help from @xtine and @donjo on that.

I was looking into using an external library like dropzone.js here, but I found myself writing almost as much code to integrate it as it took to implement with the html files api--dropzone and similar libraries are great at accepting files, but I had to fight with them to re-hydrate cached uploads from localstorage. So I opted to take the latter approach. Interested to hear whether @cmc333333 and @xtine agree--I'm not against refactoring to rely on dropzone if you all would prefer.

![35yb4ik0dq](https://cloud.githubusercontent.com/assets/1633460/14125601/1196493e-f5da-11e5-8aef-ccc71de3ecd1.gif)

cmc333333 commented on regulations/static/regulations/css/less/module/comment.less:
```diff
@@ -174,3 +174,7 @@ a.comment-index-review {
+
+input.highlight[type="file"] {
+  background: blue;
+}
```
Probably don't want this in the global namespace

jmcarp replied:
:+1:

cmc333333 commented on regulations/static/regulations/js/source/views/comment/attachment-view.js:
```diff
@@ -0,0 +1,51 @@
+  render: function() {
+    var $el = $(this.template(this.options));
+    this.options.$parent.append($el);
+    Backbone.View.prototype.setElement.call(this, $el);
```
Why use this rather than `this.setElement($el)`? Might be worth explaining in comments, at least.

jmcarp replied:
Er, no reason at all--I needed this somewhere else, but not here. Thanks for catching.

cmc333333 commented on regulations/static/regulations/js/source/views/comment/attachment-view.js:
```diff
@@ -0,0 +1,51 @@
+  handleRemove: function() {
+    CommentEvents.trigger('attachment:remove', this.options.key);
+    if (this.options.xhr) {
+      this.options.xhr.abort();
```
Nice

cmc333333 commented on regulations/static/regulations/js/source/views/comment/comment-view.js:
```diff
@@ -11,32 +11,24 @@ require('prosemirror/dist/markdown');
-    'click .queue-item': 'clearAttachment',
+    'change input[type="file"]': 'addAttachments',
+    'dragenter input[type="file"]': 'highlightDropzone',
+    'dragleave input[type="file"]': 'unhighlightDropzone',
```
Just to verify, CSS's `:hover` won't work here?

jmcarp replied:
Depends on the desired behavior. If we want to highlight the dropzone on hovering the pointer over it, then yes. This implementation is a little different, highlighting the dropzone when something's dragged over it. That was the more intuitive behavior to me, but if highlighting on hover is right, we can simplify the code a bit.

cmc333333 commented on regulations/static/regulations/js/source/views/comment/comment-view.js:
```diff
@@ -44,6 +36,7 @@ var CommentView = Backbone.View.extend({
     this.$context = this.$el.find('.comment-context');
     this.$container = this.$el.find('.editor-container');
+    this.$input = this.$el.find('input[type="file"]');
     this.$queued = this.$el.find('.queued');
```
Might be worth renaming this now, given that some of the uploads are "queued" in a different sense.

jmcarp replied:
:+1:

cmc333333 commented on regulations/static/regulations/js/source/views/comment/comment-view.js:
```diff
@@ -79,45 +74,53 @@ var CommentView = Backbone.View.extend({
+    var index = _.findIndex(this.attachmentViews, function(view) {
+      return view.options.key === key;
+    });
+    this.attachmentViews.splice(index, 1)[0].remove();
```
nbd, but is there a function in underscore that'd make this clearer?

cmc333333 commented on regulations/static/regulations/js/unittests/specs/views/comment-view-spec.js:
```diff
@@ -35,7 +36,15 @@ describe('CommentView', function() {
+    var $body = $(document.body)
+      .empty()
+      .append($el)
+      .append($template);
```
This seems suspicious -- do you really want to wipe out the body?

jmcarp replied:
Shouldn't be any harm to it, since this is just the test setup--there's nothing else in the body, and we want to clear out junk left behind by previous test runs. I can move the fixtures into a #fixtures element instead if it would be clearer.

cmc333333 replied:
Oof, sorry, forgot where I was.

cmc333333 commented on regulations/views/comment.py:
```diff
@@ -28,7 +28,7 @@ def upload_proxy(request):
         Params={
             'ContentLength': size,
-            'ContentType': 'application/octet-stream',
+            'ContentType': request.GET.get('type', 'application/octet-stream'),
```
Very helpful

cmc333333 commented:
Very cool. I think going w/o a dropzone-esque library is alright, but I defer to @xtine.

xtine commented:
I'm fine without going with a library, I would just prefer some commenting and/or documentation that explains the behavior.

jmcarp commented:
Cleaned up some issues. I added a brief description of upload flow in docs for `CommentView.addAttachment`, since that kicks off most of the work (is jsdoc cool?). If more narrative docs would be useful, I could use pointers from @xtine and @cmc333333 to let me know what should be documented.

cmc333333 commented:
:+1: from me. Will merge in a bit unless @xtine gets there first
